enhancements/monitoring/multi-tenant-alerting.md

```diff
@@ -0,0 +1,636 @@
+and where their alert notifications should be routed.
+
+### User Stories
+
+Personas:
+* Application owners: manage a project with sufficient permissions to define monitoring resources.
+* UWM admins: manage the configuration of the UWM components (edit permissions on the `openshift-user-workload-monitoring/user-workload-monitoring-config` configmap).
+* Cluster admins: manage the configuration of the Platform monitoring components.
+
+#### Story 1
+
+As an application owner, I want to use AlertmanagerConfig custom resources so
+that Alertmanager can push alert notifications for my applications to the
+receiver of my choice.
+
+#### Story 2
+
+As an application owner, I want to use AlertmanagerConfig custom
+resources so that Alertmanager can inhibit alerts based on other alerts firing
+at the same time.
+
+#### Story 3
+
+As an application owner, I want to know if my AlertmanagerConfig custom
+resource has been reconciled on the target Alertmanager so that I am confident
+that I receive alert notifications.
+
+#### Story 4
+
+As a OpenShift cluster admin, I want to allow some of my users to
+create/update/delete AlertmanagerConfig custom resources and leverage the
+platform Alertmanager cluster so that I don't have to configure alert routing
+on their behalf.
+
+#### Story 5
+
+As an OpenShift cluster admin, I want to distinguish between platform and user
+alerts so that my Alertmanager configuration can reliably handle all platform alerts.
+
+#### Story 6
+
+As an OpenShift cluster admin, I want to exclude certain user namespaces from
+modifying the Plaform Alertmanager configuration so that I can recover in case
+of breakage or bad behavior.
+
+#### Story 7
+
+As an OpenShift cluster admin, I don't want to support AlertmanagerConfig
+resources for application owners so that the configuration of the Platform
+Alertmanager cluster is completely under my control.
+
+#### Story 8
+
+As a UWM admin, I don't want to send user alerts to the Platform Alertmanager
+cluster because these alerts are managed by an external system (off-cluster Alertmanager for
+instance).
+
+### API Extensions
+
+This enhancement proposal leverages the `AlertmanagerConfig` CRD which is
+exposed by the `coreos.monitoring.com/v1alpha1` API group and defined by the
+upstream Prometheus operator.
+
+The cluster monitoring operator deploys a `ValidatingWebhookConfiguration`
+resource to check the validity of `AlertmanagerConfig` resources for things
+that can't be enforced by the OpenAPI specification. In particular, the
+AlertmanagerConfig's `Route` struct is a recursive type which isn't supported
+right now by [controller-tools][controller-tools-issue]).
+
+The validating webhook points to
+the prometheus operator's service in the `openshift-user-workload-monitoring`
+namespace (path: `/admission-alertmanagerconfigs/validate`).
+
+### Implementation Details/Notes/Constraints
+
+#### Deployment models
+
+##### Leveraging the Platform Alertmanager
```

Comment:
Inhibition is a global setting... is it possible that multi tenant AlertManager CRs could unknowingly inhibit each other?

Reply:
yes because the operator will always append a namespace matcher to the inhibition rules, e.g.

```
inhibit_rules:                                                                                      
- target_matchers:                                                                                  
  - alertname="HttpErrors"                                                                          
  source_matchers:                                                                                  
  - alertname=~"DatabaseErrors"                                                                       
  equal:                                                                                            
  - service                                                                                            
```

is turned into

```
inhibit_rules:                                                                                      
- target_matchers:                                                                                  
  - alertname="HttpErrors"
  - namespace="foo"                                                                          
  source_matchers:                                                                                  
  - alertname=~"DatabaseErrors"                                                                       
  - namespace="foo"                                                                          
  equal:                                                                                            
  - service                                                                                            
```